File: src/redirect.ts
// Redirect chain follower — manually follows redirects, recording each hop

import type { RedirectChainResult, RedirectHop, Finding } from './types';
import { USER_AGENT, FETCH_TIMEOUT } from './constants';

const MAX_HOPS = 20;

/** Follow a URL's redirect chain, recording per-hop details */
export async function followRedirects(startUrl: string): Promise<RedirectChainResult> {
  const chain: RedirectHop[] = [];
  const issues: Finding[] = [];
  const seen = new Set<string>();
  let current = startUrl;
  let loopDetected = false;
  let mixedContent = false;
  const totalStart = performance.now();

  for (let i = 0; i < MAX_HOPS; i++) {
    if (seen.has(current)) {
      loopDetected = true;
      issues.push({
        severity: 'critical',
        code: 'REDIRECT_LOOP',
        message: `Redirect loop detected: ${current} was already visited`,
        fix: 'Check your server configuration for circular redirects. A common cause is conflicting redirect rules (e.g., www → non-www and non-www → www both active).',
      });
      break;
    }
    seen.add(current);

    const hopStart = performance.now();
    let resp: Response;
    try {
      resp = await fetch(current, {
        method: 'GET',
        redirect: 'manual',
        headers: {
          'User-Agent': USER_AGENT,
          'Accept': 'text/html,application/xhtml+xml,*/*',
        },
        signal: AbortSignal.timeout(FETCH_TIMEOUT),
      });
    } catch (err) {
      issues.push({
        severity: 'critical',
        code: 'FETCH_ERROR',
        message: `Failed to fetch ${current}: ${err instanceof Error ? err.message : 'unknown error'}`,
        fix: 'The server may be down, blocking requests, or the domain may not resolve.',
      });
      break;
    }
    const hopTime = Math.round(performance.now() - hopStart);

    const location = resp.headers.get('Location');
    const status = resp.status;

    // Extract key headers for summary
    const headersSummary: Record<string, string> = {};
    for (const key of [
      'server', 'x-powered-by', 'via', 'alt-svc',
      'strict-transport-security', 'x-frame-options', 'x-content-type-options',
      'referrer-policy', 'permissions-policy', 'content-security-policy',
      'cross-origin-opener-policy', 'cross-origin-embedder-policy',
      'cross-origin-resource-policy', 'cache-control', 'content-type',
      'cf-cache-status', 'x-cache', 'age', 'vary',
    ]) {
      const val = resp.headers.get(key);
      if (val) headersSummary[key] = val;
    }

    const hop: RedirectHop = {
      url: current,
      status,
      location,
      timing_ms: hopTime,
      headers_summary: headersSummary,
      hsts: resp.headers.get('Strict-Transport-Security'),
    };
    chain.push(hop);

    // Check for mixed content (HTTPS → HTTP redirect)
    if (location) {
      const fromHttps = current.startsWith('https://');
      const toHttp = location.startsWith('http://');
      if (fromHttps && toHttp) {
        mixedContent = true;
        issues.push({
          severity: 'critical',
          code: 'MIXED_CONTENT_REDIRECT',
          message: `HTTPS → HTTP downgrade: ${current} redirects to ${location}`,
          fix: 'Never redirect from HTTPS to HTTP. Update the redirect target to use HTTPS.',
          mdn: 'https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content',
        });
      }
    }

    // Classify redirect type
    if (status === 302 || status === 303) {
      issues.push({
        severity: 'info',
        code: 'TEMPORARY_REDIRECT',
        message: `Hop ${i + 1} uses ${status} (temporary). If this is a permanent move, use 301 or 308 for better caching and SEO.`,
        fix: `Change the ${status} response to 301 (permanent) or 308 (permanent, preserves method) if the redirect is permanent.`,
      });
    }

    // Not a redirect — final destination
    if (status < 300 || status >= 400 || !location) {
      break;
    }

    // Resolve relative Location
    try {
      current = new URL(location, current).href;
    } catch {
      issues.push({
        severity: 'high',
        code: 'INVALID_LOCATION',
        message: `Invalid Location header at hop ${i + 1}: "${location}"`,
        fix: 'The Location header contains an invalid URL. Ensure it is a valid absolute or relative URL.',
      });
      break;
    }
  }

  if (chain.length >= MAX_HOPS) {
    issues.push({
      severity: 'critical',
      code: 'TOO_MANY_REDIRECTS',
      message: `Redirect chain exceeded ${MAX_HOPS} hops`,
      fix: 'Reduce the number of redirects. Most browsers give up after 20 hops.',
    });
  }

  // Check for HTTP → HTTPS upgrade (informational)
  if (chain.length >= 2) {
    const first = chain[0];
    if (first.url.startsWith('http://') && first.status >= 300 && first.status < 400 && first.location?.startsWith('https://')) {
      issues.push({
        severity: 'pass',
        code: 'HTTP_TO_HTTPS',
        message: `HTTP → HTTPS redirect in place (${first.status})`,
      });
    }
  }

  const totalTime = Math.round(performance.now() - totalStart);

  return {
    hops: chain.length,
    loop_detected: loopDetected,
    mixed_content: mixedContent,
    chain,
    issues,
    total_time_ms: totalTime,
  };
}

File: cli/redirect.go
package main

import (
	"fmt"
	"net/http"
	"net/url"
	"time"
)

const maxRedirectHops = 20

var summaryHeaders = []string{
	"server", "x-powered-by", "via", "alt-svc",
	"strict-transport-security", "x-frame-options", "x-content-type-options",
	"referrer-policy", "permissions-policy", "content-security-policy",
	"cross-origin-opener-policy", "cross-origin-embedder-policy",
	"cross-origin-resource-policy", "cache-control", "content-type",
	"cf-cache-status", "x-cache", "age", "vary",
}

// followRedirects follows a URL's redirect chain, recording per-hop details.
func followRedirects(startURL string) RedirectChain {
	chain := make([]RedirectHop, 0, 4)
	issues := make([]Finding, 0)
	seen := make(map[string]bool)
	current := startURL
	loopDetected := false
	mixedContent := false

	client := &http.Client{
		Timeout: 10 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	for i := 0; i < maxRedirectHops; i++ {
		if seen[current] {
			loopDetected = true
			issues = append(issues, Finding{
				Severity: "critical",
				Code:     "REDIRECT_LOOP",
				Message:  fmt.Sprintf("Redirect loop detected: %s was already visited", current),
				Fix:      "Check your server configuration for circular redirects. A common cause is conflicting redirect rules (e.g., www → non-www and non-www → www both active).",
			})
			break
		}
		seen[current] = true

		req, err := http.NewRequest("GET", current, nil)
		if err != nil {
			issues = append(issues, Finding{
				Severity: "critical",
				Code:     "FETCH_ERROR",
				Message:  fmt.Sprintf("Failed to build request for %s: %v", current, err),
				Fix:      "The URL may be malformed.",
			})
			break
		}
		req.Header.Set("User-Agent", fmt.Sprintf("xhttp-cli/%s", version))
		req.Header.Set("Accept", "text/html,application/xhtml+xml,*/*")

		hopStart := time.Now()
		resp, err := client.Do(req)
		hopTime := float64(time.Since(hopStart).Milliseconds())

		if err != nil {
			issues = append(issues, Finding{
				Severity: "critical",
				Code:     "FETCH_ERROR",
				Message:  fmt.Sprintf("Failed to fetch %s: %v", current, err),
				Fix:      "The server may be down, blocking requests, or the domain may not resolve.",
			})
			break
		}
		resp.Body.Close()

		location := resp.Header.Get("Location")
		status := resp.StatusCode

		// Extract key headers for summary
		headersSummary := make(map[string]string)
		for _, key := range summaryHeaders {
			if val := resp.Header.Get(key); val != "" {
				headersSummary[key] = val
			}
		}

		hop := RedirectHop{
			URL:            current,
			Status:         status,
			Location:       location,
			TimingMs:       hopTime,
			HeadersSummary: headersSummary,
		}
		chain = append(chain, hop)

		// Check for mixed content (HTTPS → HTTP redirect)
		if location != "" {
			if len(current) >= 8 && current[:8] == "https://" && len(location) >= 7 && location[:7] == "http://" {
				mixedContent = true
				issues = append(issues, Finding{
					Severity: "critical",
					Code:     "MIXED_CONTENT_REDIRECT",
					Message:  fmt.Sprintf("HTTPS → HTTP downgrade: %s redirects to %s", current, location),
					Fix:      "Never redirect from HTTPS to HTTP. Update the redirect target to use HTTPS.",
					MDN:      "https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content",
				})
			}
		}

		// Classify redirect type
		if status == 302 || status == 303 {
			issues = append(issues, Finding{
				Severity: "info",
				Code:     "TEMPORARY_REDIRECT",
				Message:  fmt.Sprintf("Hop %d uses %d (temporary). If this is a permanent move, use 301 or 308 for better caching and SEO.", i+1, status),
				Fix:      fmt.Sprintf("Change the %d response to 301 (permanent) or 308 (permanent, preserves method) if the redirect is permanent.", status),
			})
		}

		// Not a redirect — final destination
		if status < 300 || status >= 400 || location == "" {
			break
		}

		// Resolve relative Location
		parsed, err := url.Parse(location)
		if err != nil {
			issues = append(issues, Finding{
				Severity: "high",
				Code:     "INVALID_LOCATION",
				Message:  fmt.Sprintf("Invalid Location header at hop %d: %q", i+1, location),
				Fix:      "The Location header contains an invalid URL. Ensure it is a valid absolute or relative URL.",
			})
			break
		}
		base, _ := url.Parse(current)
		current = base.ResolveReference(parsed).String()
	}

	if len(chain) >= maxRedirectHops {
		issues = append(issues, Finding{
			Severity: "critical",
			Code:     "TOO_MANY_REDIRECTS",
			Message:  fmt.Sprintf("Redirect chain exceeded %d hops", maxRedirectHops),
			Fix:      "Reduce the number of redirects. Most browsers give up after 20 hops.",
		})
	}

	// Check for HTTP → HTTPS upgrade (informational)
	if len(chain) >= 2 {
		first := chain[0]
		if len(first.URL) >= 7 && first.URL[:7] == "http://" &&
			first.Status >= 300 && first.Status < 400 &&
			first.Location != "" && len(first.Location) >= 8 && first.Location[:8] == "https://" {
			issues = append(issues, Finding{
				Severity: "pass",
				Code:     "HTTP_TO_HTTPS",
				Message:  fmt.Sprintf("HTTP → HTTPS redirect in place (%d)", first.Status),
			})
		}
	}

	return RedirectChain{
		Hops:         len(chain),
		LoopDetected: loopDetected,
		MixedContent: mixedContent,
		Chain:        chain,
		Issues:       issues,
	}
}
